### PR TITLE
[Octave] New builder - v9.2.0

### PR DESCRIPTION
**Pull request by ViralBShah**
Addressing #8397

### CODE REVIEW

**Comment by ViralBShah** [comment]
These are all failing with "No space left on device". @giordano Any ideas?

**Comment by giordano** [comment]
Disk is good, 77% of utilization.  What directory is being filled?  I think `/tmp` relatively small.

File: O/Octave/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,60 @@
+using BinaryBuilder, Pkg
+
+name = "Octave"
+version = v"9.2.0"
+
+# Collection of sources required to build Octave
+sources = [
+   ArchiveSource("https://ftpmirror.gnu.org/octave/octave-9.2.0.tar.gz",
```
Comment (giordano) [inline O/Octave/build_tarballs.jl:8]:
```suggestion
   ArchiveSource("https://ftpmirror.gnu.org/octave/octave-$(version).tar.gz",
```

File: O/Octave/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+# Base configure flags
+FLAGS=(
+    --prefix="$prefix"
+    --build=${MACHTYPE}
+    --host="${target}"
+    --enable-shared
+    --disable-static
+    --with-blas="-L${prefix}/lib -l${LBT}"
+    --with-lapack="-L${prefix}/lib -l${LBT}"
```
Comment (giordano) [inline O/Octave/build_tarballs.jl:33]:
```suggestion
    --with-blas="-L${libdir} -l${LBT}"
    --with-lapack="-L${libdir} -l${LBT}"
```

File: O/Octave/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                  "0636554b05996997e431caad4422c00386d2d7c68900472700fecf5ffeb7c991"),
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/octave*
+
+export CPPFLAGS="-I${includedir}"
+export LDFLAGS="-L${libdir}"
+
```
Comment (giordano) [inline O/Octave/build_tarballs.jl:18]: Are these necessary?  I think they may be already in the compiler search paths.

File: O/Octave/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+platforms = supported_platforms()
+platforms = expand_cxxstring_abis(platforms)
+
+# The products that we will ensure are always built
+products = [
+    ExecutableProduct("octave", :octave),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
```
Comment (giordano) [inline O/Octave/build_tarballs.jl:51]:
```suggestion
dependencies =[
```

File: O/Octave/build_tarballs.jl (hunk 3 above)
Comment (ViralBShah) [reply O/Octave/build_tarballs.jl:18]: Just copied from another file somewhere to get started. I was having trouble with BLAS and LAPACK libraries which it wasn't finding. Let me remove and try again.

**Comment by ViralBShah** [comment]
> Disk is good, 77% of utilization. What directory is being filled? I think `/tmp` relatively small.

All the builds just stopped at the same point in the compilation, which felt a little weird. Let's see what happens with the new builds just going on.

**Comment by giordano** [comment]
They seem to be writing in /tmp, no?

**Comment by ViralBShah** [comment]
> They seem to be writing in /tmp, no?

I can't quite tell - but now a bunch of builds are going through.

**Comment by giordano** [comment]
> I can't quite tell

https://buildkite.com/julialang/yggdrasil/builds/14405#0192f38a-003e-4090-a710-fdda681c048f/6-11202

```
[19:48:25] libinterp/octave-value/ov-uint32.cc:79:64: fatal error: error writing to /tmp/ccHiPhhD.s: No space left on device
```

Looks quite clear to me.

**Comment by ViralBShah** [comment]
> https://buildkite.com/julialang/yggdrasil/builds/14405#0192f38a-003e-4090-a710-fdda681c048f/6-11202
> 
> ```
> [19:48:25] libinterp/octave-value/ov-uint32.cc:79:64: fatal error: error writing to /tmp/ccHiPhhD.s: No space left on device
> ```
> 
> Looks quite clear to me.

Yes, so some of these are still running out of space in `/tmp`. Is there an easy fix?

**Comment by giordano** [comment]
I don't know, it depends on how the build system decides to pick the `/tmp` directory: it could be hardcoded, it could be using an environment variable (`TMP`  or `TMPDIR` are common options), it could be an option to the build system. I haven't looked at the code, I don't know what they're doing.

**Comment by ViralBShah** [comment]
@giordano Thanks. I didn't know about TMPDIR. I am hoping it works since it seems a bit standard for autoconf systems. I'll merge this later today.

**Comment by ViralBShah** [comment]
@ararslan FreeBSD is failing with this. I wonder if there is an easy known fix. Just checking before I disable it.

`error: redefinition of 'octave_interrupt_state' with a different type: 'atomic<sig_atomic_t>' vs 'atomic<int>'`

**Comment by ViralBShah** [comment]
This should probably fail the build, but it isn't. Not sure what is causing it, but feels like we need to fix it before merging:

```
2024-11-04 17:19:11 EST | [ Info: Checking shared library lib/octave/9.2.0/liboctave.so.11.0.1
-- | --
  | 2024-11-04 17:19:11 EST |  
  | 2024-11-04 17:19:11 EST | signal (11): Segmentation fault
  | 2024-11-04 17:19:11 EST | in expression starting at none:2
  | 2024-11-04 17:19:11 EST | unknown function (ip: (nil))
  | 2024-11-04 17:19:11 EST | Allocations: 2722 (Pool: 2712; Big: 10); GC: 0
  | 2024-11-04 17:19:11 EST | [ Info: lib/octave/9.2.0/liboctave.so.11.0.1 already has SONAME "liboctave.so.11"
  | 2024-11-04 17:19:11 EST | [ Info: Checking shared library lib/octave/9.2.0/liboctinterp.so.12.0.1
  | 2024-11-04 17:19:12 EST | ERROR: could not load library "/cache/build/yggy-amdci7-15/julialang/yggdrasil/O/Octave/build/x86_64-linux-gnu-cxx11/VnRTm55s/x86_64-linux-gnu-libgfortran5-cxx11/destdir/lib/octave/9.2.0/liboctinterp.so.12.0.1"
  | 2024-11-04 17:19:12 EST | /cache/julia-buildkite-plugin/julia_installs/bin/linux/x64/1.7/julia-1.7-latest-linux-x86_64/bin/../lib/julia/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /cache/build/yggy-amdci7-15/julialang/yggdrasil/O/Octave/build/x86_64-linux-gnu-cxx11/VnRTm55s/x86_64-linux-gnu-libgfortran5-cxx11/destdir/lib/octave/9.2.0/liboctinterp.so.12.0.1)
  | 2024-11-04 17:19:12 EST | ┌ Warning: lib/octave/9.2.0/liboctinterp.so.12.0.1 cannot be dlopen()'ed
```

In general this recipe will be tough to line up all the 64-bit stuff. It is unclear whether Octave can use the suffixed versions of BLAS and LAPACK the way we do. 

We could build a 32-bit version, but it needs things like Arpack and SuiteSparse to have 32-bit versions in Yggdrasil, which we do not have at the moment.

@Krastanov Any thoughts on if this would all be valuable to chase through?

File: O/Octave/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    Dependency("Sundials_jll"),
+    Dependency("HDF5_jll"),
+    Dependency("rapidjson_jll"),
+    Dependency("libsndfile_jll"),
+    Dependency("GraphicsMagick_jll"),
+]
+
+# Build the tarballs.
+build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies;
+               julia_compat="1.10", clang_use_lld=false, preferred_gcc_version=v"12")
```
Comment (giordano) [inline O/Octave/build_tarballs.jl:85]: Gcc 12 is too new for the audit with Julia v1.7

**Comment by ararslan** [comment]
Looks like FreeBSD applies these patches for Octave: https://github.com/freebsd/freebsd-ports/tree/main/math/octave/files, one of which looks like it would fix the int thing

File: O/Octave/build_tarballs.jl (hunk 5 above)
Comment (ViralBShah) [reply O/Octave/build_tarballs.jl:85]: Which one should I use? GCC 9 was definitely having issues, IIRC. Maybe 10?

Comment (giordano) [reply O/Octave/build_tarballs.jl:85]: 10 should be fine for v1.6 (and v1.7)

**Pull request merged**